File: bot_engine/memory/extractor.py
"""Auto-extract facts, decisions, and preferences from conversations using Claude CLI."""

from __future__ import annotations

import json
import logging
import os
import re
import subprocess
import uuid

import psycopg2

from bot_engine.memory.opensearch_client import OpenSearchClient
from bot_engine.memory.store import MessageStore

log = logging.getLogger(__name__)

CLAUDE_CLI_PATH = os.environ.get("CLAUDE_CLI_PATH", "claude")

EXTRACTION_PROMPT_TEMPLATE = """Analyze the following conversation and extract important information that should be remembered for future conversations.

Extract ONLY genuinely important items in these categories:
- **fact**: Factual information about the project, user, or system
- **decision**: Decisions made during the conversation
- **preference**: User preferences about how things should be done

Important extraction rules:
- Prefer durable behavior preferences over one-off task requests.
- If the user gives a short imperative follow-up like "해", "계속 해", "이어서 해", or "보강해", use the surrounding conversation to rewrite it into a clear standalone preference when it reinforces an ongoing instruction.
- When the user asks the assistant to keep checking, keep improving, or continue without waiting, extract that as a **preference**.
- Keep each memory self-contained and specific enough to reuse later.

Conversation:
{conversation}

Respond with a JSON array ONLY. No other text. Example:
[
  {{"category": "fact", "content": "사용자는 Airflow 3.1.8을 사용 중"}},
  {{"category": "decision", "content": "마스터 봇은 sonnet 모델을 사용하기로 결정"}},
  {{"category": "preference", "content": "사용자는 간결한 답변을 선호"}}
]

If nothing important to extract, respond with: []"""

_CONTINUE_PATTERNS = (
    "계속 해",
    "계속해",
    "이어서 해",
    "이어서해",
    "계속 보강",
    "보강해",
    "보강 할건 계속 보강해",
    "보강할건 계속 보강해",
)
_PROACTIVE_PATTERNS = (
    "기다리지",
    "대기하지",
    "먼저 말",
    "이어서 말",
    "추가로 얘기",
    "계속 점검",
    "끝까지 점검",
)


class MemoryExtractor:
    """Extract and store memories from conversations."""

    def __init__(self) -> None:
        self._dsn = os.environ.get("DATABASE_DSN", "")
        self._store = MessageStore()
        self._os_client: OpenSearchClient | None = None
        try:
            self._os_client = OpenSearchClient()
        except Exception as e:
            log.warning("OpenSearch client init failed: %s", e)

    def extract_and_store(
        self, bot_id: str, user_id: str, conversation: list[dict]
    ) -> int:
        """Extract memories from conversation and store in PostgreSQL + OpenSearch.

        Returns number of memories extracted.
        """
        if not conversation or len(conversation) < 2:
            return 0

        # Build conversation text from last 10 messages; truncate each to 500 chars
        conv_lines = []
        for msg in conversation[-10:]:
            role = "User" if msg.get("role") == "user" else "Assistant"
            content = msg.get("content", "")[:500]
            conv_lines.append(f"{role}: {content}")
        conv_text = "\n".join(conv_lines)

        prompt = EXTRACTION_PROMPT_TEMPLATE.format(conversation=conv_text)
        extracted = self._call_claude(prompt)
        if not extracted:
            return 0

        memories = self._parse_extraction(extracted)
        memories.extend(self._infer_preferences(conversation))
        if not memories:
            return 0

        stored = 0
        for mem in memories:
            try:
                memory_id = str(uuid.uuid4())
                category = mem.get("category", "fact")
                content = re.sub(r"\s+", " ", mem.get("content", "")).strip()
                if not content:
                    continue
                if self._store.memory_exists(bot_id, category, content):
                    continue

                if self._store_pg(memory_id, bot_id, user_id, category, content):
                    if self._os_client:
                        self._os_client.index_memory(
                            memory_id=memory_id,
                            bot_id=bot_id,
                            content=content,
                            category=category,
                            user_id=user_id,
                        )
                    stored += 1
            except Exception as e:
                log.warning("Failed to store memory: %s", e)

        if stored:
            log.info(
                "Extracted and stored %d memories for bot %s", stored, bot_id
            )
        return stored

    def _infer_preferences(self, conversation: list[dict]) -> list[dict]:
        """Derive durable preferences from short imperative user follow-ups."""
        user_messages = [
            re.sub(r"\s+", " ", (msg.get("content") or "")).strip()
            for msg in conversation
            if msg.get("role") == "user" and (msg.get("content") or "").strip()
        ]
        if not user_messages:
            return []

        joined = " ".join(user_messages).lower()
        inferred: list[dict] = []

        if any(pattern in joined for pattern in _CONTINUE_PATTERNS):
            inferred.append(
                {
                    "category": "preference",
                    "content": (
                        "사용자는 개선이나 보강 요청을 시작한 뒤에는 멈추지 말고 계속 이어서 보강하기를 원함"
                    ),
                }
            )

        if any(pattern in joined for pattern in _PROACTIVE_PATTERNS):
            inferred.append(
                {
                    "category": "preference",
                    "content": (
                        "사용자는 확인이나 점검 요청 후 추가로 볼 것이 있으면 기다리지 말고 먼저 이어서 보고하기를 원함"
                    ),
                }
            )

        # Extremely short follow-ups like "해" are only meaningful when the
        # surrounding conversation already indicates a continue/keep-going intent.
        if user_messages[-1] in {"해", "해줘", "해봐"} and any(
            pattern in joined for pattern in _CONTINUE_PATTERNS
        ):
            inferred.append(
                {
                    "category": "preference",
                    "content": (
                        "사용자는 짧은 후속 지시라도 직전 맥락을 이어받아 작업을 계속 진행하기를 원함"
                    ),
                }
            )

        deduped: list[dict] = []
        seen: set[tuple[str, str]] = set()
        for item in inferred:
            key = (item["category"], item["content"])
            if key in seen:
                continue
            seen.add(key)
            deduped.append(item)
        return deduped

    def _call_claude(self, prompt: str) -> str | None:
        """Call Claude CLI (haiku) for extraction.

        Returns raw output string or None on failure.
        """
        try:
            env = {
                **os.environ,
                "HOME": os.environ.get("CONTAINER_HOME", "/home/appuser"),
                "NO_COLOR": "1",
            }
            result = subprocess.run(
                [
                    CLAUDE_CLI_PATH, "-p", prompt,
                    "--model", "haiku",
                    "--max-turns", "1",
                ],
                env=env,
                capture_output=True,
                text=True,
                timeout=30,
                stdin=subprocess.DEVNULL,
            )
            if result.returncode != 0:
                return None
            return result.stdout.strip() or None
        except Exception as e:
            log.warning("Claude CLI extraction failed: %s", e)
            return None

    def _parse_extraction(self, text: str) -> list[dict]:
        """Parse JSON array from Claude's response.

        Handles plain JSON and markdown code blocks.
        """
        text = text.strip()
        if "```" in text:
            match = re.search(r"```(?:json)?\s*\n?(.*?)\n?\s*```", text, re.DOTALL)
            if match:
                text = match.group(1).strip()
        try:
            parsed = json.loads(text)
            if isinstance(parsed, list):
                return [m for m in parsed if isinstance(m, dict) and "content" in m]
        except json.JSONDecodeError:
            pass
        return []

    def _store_pg(
        self,
        memory_id: str,
        bot_id: str,
        user_id: str,
        category: str,
        content: str,
    ) -> bool:
        """Store memory in PostgreSQL.

        Uses a zero-vector placeholder for the embedding column until
        Voyage API is available.
        """
        if not self._dsn:
            return False
        try:
            conn = psycopg2.connect(self._dsn)
            with conn.cursor() as cur:
                zero_vec = "[" + ",".join(["0"] * 1024) + "]"
                cur.execute(
                    """INSERT INTO memories
                        (id, bot_id, user_id, category, content, embedding)
                        VALUES (%s, %s, %s, %s, %s, %s::vector)""",
                    (memory_id, bot_id, user_id, category, content, zero_vec),
                )
            conn.commit()
            conn.close()
            return True
        except Exception as e:
            log.warning("Failed to store memory in PostgreSQL: %s", e)
            return False

File: go/internal/memory/extractor.go
package memory

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"regexp"
	"strings"
	"unicode/utf8"

	"github.com/baekenough/customclaw/internal/llm"
)

const (
	maxMessages      = 10
	maxMessageRunes  = 500
	extractionPrompt = `Analyze the following conversation and extract important information.
Extract ONLY genuinely important items in these categories:
- fact: Factual information about the project, user, or system
- decision: Decisions made during the conversation
- preference: User preferences about how things should be done
- action: Actions taken or requested during the conversation
- context: Important contextual information for future reference

Rules:
- Skip trivial exchanges, greetings, and filler content
- Each item must be self-contained and meaningful without the full conversation
- Be concise but complete
- Only extract items that would genuinely help in future conversations

Respond with a JSON array ONLY. No markdown, no explanation. Format:
[{"category": "fact", "content": "..."}, {"category": "decision", "content": "..."}]

If nothing important was discussed, respond with: []

Conversation:
`
)

// continuePatterns matches Korean phrases indicating the user wants work to continue.
var continuePatterns = []*regexp.Regexp{
	regexp.MustCompile(`계속\s*해`),
	regexp.MustCompile(`이어서\s*해`),
	regexp.MustCompile(`보강해`),
	regexp.MustCompile(`계속\s*진행`),
	regexp.MustCompile(`계속\s*하`),
}

// proactivePatterns matches Korean phrases indicating the user wants proactive responses.
var proactivePatterns = []*regexp.Regexp{
	regexp.MustCompile(`기다리지`),
	regexp.MustCompile(`대기하지`),
	regexp.MustCompile(`먼저\s*말`),
	regexp.MustCompile(`알아서`),
	regexp.MustCompile(`먼저\s*해`),
}

// followUpRe detects short follow-up requests of 5 chars or fewer.
var followUpRe = regexp.MustCompile(`^.{1,5}$`)

// extractedMemory is the JSON structure returned by the LLM.
type extractedMemory struct {
	Category string `json:"category"`
	Content  string `json:"content"`
}

// jsonArrayRe strips optional markdown code-fence wrapping from an LLM response.
var jsonArrayRe = regexp.MustCompile("(?s)```(?:json)?\\s*(\\[.*?\\])\\s*```")

// MemoryExtractor distils conversation history into long-term memory entries.
// It calls the LLM with a haiku-class model for cost efficiency and stores
// extracted facts in PostgreSQL, optionally indexing them in OpenSearch.
type MemoryExtractor struct {
	provider llm.Provider
	store    *MessageStore
	osClient *OpenSearchClient
	dsn      string
}

// NewMemoryExtractor creates a MemoryExtractor with the supplied dependencies.
// osClient may be nil when OpenSearch is not configured.
func NewMemoryExtractor(
	provider llm.Provider,
	store *MessageStore,
	osClient *OpenSearchClient,
	dsn string,
) *MemoryExtractor {
	return &MemoryExtractor{
		provider: provider,
		store:    store,
		osClient: osClient,
		dsn:      dsn,
	}
}

// ExtractAndStore analyses the recent conversation for the given bot and channel
// and persists any extracted facts to the memory store.
// All errors are logged at Warn level; the method never blocks message processing.
func (e *MemoryExtractor) ExtractAndStore(ctx context.Context, botID, channelID string) error {
	hKey := "channel:" + channelID

	history, err := e.store.GetHistory(ctx, hKey, maxMessages)
	if err != nil {
		slog.Warn("extractor: failed to load history",
			"bot", botID,
			"channel", channelID,
			"error", err,
		)
		return nil
	}
	if len(history) == 0 {
		return nil
	}

	convText := buildConversationText(history)

	extracted, extractErr := e.callLLM(ctx, convText)
	if extractErr != nil {
		slog.Warn("extractor: llm call failed",
			"bot", botID,
			"channel", channelID,
			"error", extractErr,
		)
	}

	inferred := inferPreferences(history)
	all := append(extracted, inferred...)

	if len(all) == 0 {
		return nil
	}

	stored := 0
	for _, m := range all {
		if m.Content == "" {
			continue
		}
		exists, checkErr := e.store.MemoryExists(ctx, botID, m.Category, m.Content)
		if checkErr != nil {
			slog.Warn("extractor: duplicate check failed", "error", checkErr)
		}
		if exists {
			continue
		}
		if storeErr := e.storeMemory(ctx, botID, channelID, m); storeErr != nil {
			slog.Warn("extractor: failed to store memory",
				"category", m.Category,
				"error", storeErr,
			)
			continue
		}
		stored++
	}

	slog.Info("extractor: memories stored",
		"bot", botID,
		"channel", channelID,
		"count", stored,
	)
	return nil
}

// callLLM sends the conversation text to the LLM and parses the JSON response.
func (e *MemoryExtractor) callLLM(ctx context.Context, convText string) ([]extractedMemory, error) {
	req := &llm.Request{
		UserMessage: extractionPrompt + convText,
		Model:       "haiku",
		MaxTurns:    1,
	}

	resp, err := e.provider.Complete(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("llm complete: %w", err)
	}

	return parseExtraction(resp.Text), nil
}

// storeMemory persists a single extracted memory to PostgreSQL and optionally OpenSearch.
func (e *MemoryExtractor) storeMemory(ctx context.Context, botID, userID string, m extractedMemory) error {
	memoryID, err := e.store.StoreMemory(ctx, botID, userID, m.Category, m.Content)
	if err != nil {
		return err
	}
	if e.osClient != nil {
		if err := e.osClient.IndexMemory(ctx, memoryID, botID, m.Content, m.Category, userID); err != nil {
			slog.Warn("extractor: opensearch index failed", "error", err)
			// Non-fatal: PostgreSQL is the source of truth.
		}
	}
	return nil
}

// buildConversationText formats the last maxMessages messages into a prompt string.
// Each message content is truncated to maxMessageRunes runes.
// Empty messages are filtered.
func buildConversationText(messages []Message) string {
	start := 0
	if len(messages) > maxMessages {
		start = len(messages) - maxMessages
	}

	var sb strings.Builder
	for _, m := range messages[start:] {
		content := strings.TrimSpace(m.Content)
		if content == "" {
			continue
		}
		if utf8.RuneCountInString(content) > maxMessageRunes {
			runes := []rune(content)
			content = string(runes[:maxMessageRunes])
		}
		role := m.Role
		if role == "user" {
			role = "User"
		} else {
			role = "Assistant"
		}
		sb.WriteString(role)
		sb.WriteString(": ")
		sb.WriteString(content)
		sb.WriteString("\n")
	}
	return sb.String()
}

// parseExtraction decodes the LLM JSON response into a slice of extractedMemory.
// It handles optional markdown code-fence wrapping and filters entries with no content.
// Returns an empty slice on any parse error.
func parseExtraction(raw string) []extractedMemory {
	text := strings.TrimSpace(raw)

	// Strip markdown code fences if present.
	if m := jsonArrayRe.FindStringSubmatch(text); len(m) == 2 {
		text = m[1]
	}

	// Find the JSON array bounds directly for cases without code fences.
	start := strings.Index(text, "[")
	end := strings.LastIndex(text, "]")
	if start == -1 || end == -1 || end <= start {
		return nil
	}
	text = text[start : end+1]

	var memories []extractedMemory
	if err := json.Unmarshal([]byte(text), &memories); err != nil {
		return nil
	}

	// Filter entries that lack content.
	result := memories[:0]
	for _, m := range memories {
		if strings.TrimSpace(m.Content) != "" {
			result = append(result, m)
		}
	}
	return result
}

// inferPreferences analyses the conversation for pattern-based preference signals
// and returns synthetic memory entries for any patterns found.
// Each preference type is emitted at most once per conversation.
func inferPreferences(messages []Message) []extractedMemory {
	if len(messages) == 0 {
		return nil
	}

	// Combine all user messages into a single string for pattern matching.
	var userText strings.Builder
	for _, m := range messages {
		if m.Role == "user" {
			userText.WriteString(m.Content)
			userText.WriteString("\n")
		}
	}
	combined := userText.String()

	var result []extractedMemory
	seen := make(map[string]bool)

	for _, p := range continuePatterns {
		if p.MatchString(combined) {
			key := "continue"
			if !seen[key] {
				seen[key] = true
				result = append(result, extractedMemory{
					Category: "preference",
					Content:  "User prefers that work continues without waiting for confirmation",
				})
			}
			break
		}
	}

	for _, p := range proactivePatterns {
		if p.MatchString(combined) {
			key := "proactive"
			if !seen[key] {
				seen[key] = true
				result = append(result, extractedMemory{
					Category: "preference",
					Content:  "User wants proactive responses without being asked first",
				})
			}
			break
		}
	}

	// Detect short follow-up messages (e.g. "해", "응", "맞아") that appear after
	// a substantial assistant response — a signal the user prefers continuations.
	if len(messages) >= 2 {
		last := messages[len(messages)-1]
		prev := messages[len(messages)-2]
		if last.Role == "user" &&
			prev.Role == "assistant" &&
			followUpRe.MatchString(strings.TrimSpace(last.Content)) {
			key := "followup"
			if !seen[key] {
				seen[key] = true
				result = append(result, extractedMemory{
					Category: "preference",
					Content:  "User often sends short follow-up messages to continue work",
				})
			}
		}
	}

	return result
}
